[19.0][MIG] sale_triple_discount: Migration to 19.0

lef-adhoc opened the pull request:
Based on this pr: https://github.com/OCA/sale-workflow/pull/3889

Depends on:

- [x] <https://github.com/OCA/account-invoicing/pull/2144>

tafaRU committed:
[ADD] sale_triple_discount

tafaRU committed:
[ADD] sale_triple_discount

chienandalu committed:
[MIG] sale_triple_discount: Migration to 10.0

OCA Transbot updated translations from Transifex

[UPD] Update sale_triple_discount.pot

chienandalu committed:
[MIG] sale_triple_discount: Migration to 10.0

OCA Transbot updated translations from Transifex

[UPD] Update sale_triple_discount.pot

thomaspot committed:
Translated using Weblate (Nederlands (nl_NL))

Currently translated at 100,0% (5 of 5 strings)

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_triple_discount/nl_NL/

thomaspot committed:
Translated using Weblate (Nederlands (nl_NL))

Currently translated at 100,0% (5 of 5 strings)

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_triple_discount/nl_NL/

thomaspot committed:
Translated using Weblate (Dutch (nl_NL))

Currently translated at 100.0% (5 of 5 strings)

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_triple_discount/nl_NL/

thomaspot committed:
Translated using Weblate (Dutch (nl_NL))

Currently translated at 100.0% (5 of 5 strings)

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_triple_discount/nl_NL/

mpanarin committed:
[IMP] 10.0 sale_triple_discount: rework discount computation

mpanarin committed:
[IMP] 10.0 sale_triple_discount: rework discount computation

jbaudoux committed:
sale_triple_discount: Fix tax total on report: include the discount2/3 in the computation

jbaudoux committed:
sale_triple_discount: Fix tax total on report: include the discount2/3 in the computation

SimoRubi committed:
[FIX] sale_triple_discount: Correctly compute taxes with round_globally configuration

SimoRubi committed:
[FIX] sale_triple_discount: Correctly compute taxes with round_globally configuration

SimoRubi committed:
[FIX] sale_triple_discount: Store values in cache for recomputation consistency

[FIX] sale_triple_discount: Invalidating the cache of specific records before recomputing price and discount

[UPD] Update sale_triple_discount.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_triple_discount/

SimoRubi committed:
[FIX] sale_triple_discount: Store values in cache for recomputation consistency

[FIX] sale_triple_discount: Invalidating the cache of specific records before recomputing price and discount

[UPD] Update sale_triple_discount.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_triple_discount/

grindtildeath committed:
Play onchange after invoice creation to ensure discounts are applied

grindtildeath committed:
Play onchange after invoice creation to ensure discounts are applied

grindtildeath committed:
Fix groups on sale.order view

[UPD] Update sale_triple_discount.pot

grindtildeath committed:
Fix groups on sale.order view

[UPD] Update sale_triple_discount.pot

tafaRU committed:
[MIG] sale_triple_discount: Migration to 12.0

[UPD] Update sale_triple_discount.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_triple_discount/

tafaRU committed:
[MIG] sale_triple_discount: Migration to 12.0

[UPD] Update sale_triple_discount.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_triple_discount/

ps-tubtim committed:
[IMP] sale_triple_discount: black, isort, prettier

ps-tubtim committed:
[IMP] sale_triple_discount: black, isort, prettier

ps-tubtim committed:
[MIG] sale_triple_discount: Migration to 13.0

ps-tubtim committed:
[MIG] sale_triple_discount: Migration to 13.0

ferran-S73 committed:
[IMP] sale_triple_discount: black, isort, prettier

ferran-S73 committed:
[IMP] sale_triple_discount: black, isort, prettier

ferran-S73 committed:
[14.0][MIG] sale_triple_discount: Migration to V14

ferran-S73 committed:
[14.0][MIG] sale_triple_discount: Migration to V14

ferran-S73 committed:
[14.0][IMP] sale_triple_discount: add disc. fields in sale report templ.

ferran-S73 committed:
[14.0][IMP] sale_triple_discount: add disc. fields in sale report templ.

ferran-S73 committed:
[14.0] sale_triple_discount: sale_order_line_discount_validation compat.

[UPD] Update sale_triple_discount.pot

[UPD] README.rst

ferran-S73 committed:
[14.0] sale_triple_discount: sale_order_line_discount_validation compat.

[UPD] Update sale_triple_discount.pot

[UPD] README.rst

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (11 of 11 strings)

Translation: sale-workflow-14.0/sale-workflow-14.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-14-0/sale-workflow-14-0-sale_triple_discount/it/

[UPD] README.rst

[UPD] README.rst

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (11 of 11 strings)

Translation: sale-workflow-14.0/sale-workflow-14.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-14-0/sale-workflow-14-0-sale_triple_discount/it/

[UPD] README.rst

[UPD] README.rst

klodr committed:
Translated using Weblate (French)

Currently translated at 0.0% (0 of 11 strings)

Translation: sale-workflow-14.0/sale-workflow-14.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-14-0/sale-workflow-14-0-sale_triple_discount/fr/

klodr committed:
Translated using Weblate (French)

Currently translated at 0.0% (0 of 11 strings)

Translation: sale-workflow-14.0/sale-workflow-14.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-14-0/sale-workflow-14-0-sale_triple_discount/fr/

manuelregidor committed:
[IMP] sale_triple_discount: black, isort, prettier

manuelregidor committed:
[IMP] sale_triple_discount: black, isort, prettier

manuelregidor committed:
[MIG]sale_triple_discount: Migration to 15.0

[UPD] Update sale_triple_discount.pot

[UPD] README.rst

manuelregidor committed:
[MIG]sale_triple_discount: Migration to 15.0

[UPD] Update sale_triple_discount.pot

[UPD] README.rst

OCA-git-bot committed:
sale_triple_discount 15.0.1.0.1

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_triple_discount/

OCA-git-bot committed:
sale_triple_discount 15.0.1.0.1

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_triple_discount/

sbejaoui committed:
[16.0][MIG] - sale_triple_discount

sbejaoui committed:
[16.0][MIG] - sale_triple_discount

Commit:
[UPD] Update sale_triple_discount.pot

Commit:
[UPD] Update sale_triple_discount.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
sale_triple_discount 16.0.1.0.1

OCA-git-bot committed:
sale_triple_discount 16.0.1.0.1

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/

alan196 committed:
Added translation using Weblate (Spanish (Mexico))

alan196 committed:
Added translation using Weblate (Spanish (Mexico))

alan196 committed:
Translated using Weblate (Spanish (Mexico))

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/es_MX/

alan196 committed:
Translated using Weblate (Spanish (Mexico))

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/es_MX/

Commit:
Translated using Weblate (Italian)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/it/

Commit:
Translated using Weblate (Italian)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/it/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/es/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/es/

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/it/

renda-dev committed:
[IMP] sale_triple_discount: improve module extensibility

renda-dev committed:
[IMP] sale_triple_discount: improve module extensibility

jappi00 committed:
Translated using Weblate (German)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/de/

jappi00 committed:
Translated using Weblate (German)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/de/

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
[BOT] post-merge updates

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/pt_BR/

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/pt_BR/

lmignon committed:
[FIX] sale_triple_discount: Remove call to deprecated

Method 'invalidate_cache' is deprecated. To invalidate specific field's value into the cache for a recordset we must use the 'invalidate_recordset' method on the recordset itself.

lmignon committed:
[FIX] sale_triple_discount: Remove call to deprecated

Method 'invalidate_cache' is deprecated. To invalidate specific field's value into the cache for a recordset we must use the 'invalidate_recordset' method on the recordset itself.

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
[BOT] post-merge updates

lmignon committed:
[IMP] sale_triple_discount: Improves perfs and consistency

When computing the amount total on a sale.order.line, we now avoid to play with the cache since this change introduced in https://github.com/OCA/sale-workflow/commit/e3e59ba15ebcf1f91124eda60a694c0a76fabb69 introduces a decline in performance. When the cache in invalidated, fields are flushed to the database and depending computed fields are recomputed..... This ultimately leads to temporary inconsistencies breaking others addons relaying on this one. Moreover, the method used in this original commit 'self.invalidate_cache(fnames=self._discount_fields(), ids=self.ids)' is deprecated and the use of the recommended new one lead to others side effects. The original approach is replaced by a contextual method modifying only the discount field when we need it's needed to don't break code relaying on the fact that the discount field is the total discount to apply to a line. This contextual method ensures that the original value is properly restored on exit and that it will not trigger additional recompute due to the temporary change of the discount value

lmignon committed:
[IMP] sale_triple_discount: Improves perfs and consistency

When computing the amount total on a sale.order.line, we now avoid to play with the cache since this change introduced in https://github.com/OCA/sale-workflow/commit/e3e59ba15ebcf1f91124eda60a694c0a76fabb69 introduces a decline in performance. When the cache in invalidated, fields are flushed to the database and depending computed fields are recomputed..... This ultimately leads to temporary inconsistencies breaking others addons relaying on this one. Moreover, the method used in this original commit 'self.invalidate_cache(fnames=self._discount_fields(), ids=self.ids)' is deprecated and the use of the recommended new one lead to others side effects. The original approach is replaced by a contextual method modifying only the discount field when we need it's needed to don't break code relaying on the fact that the discount field is the total discount to apply to a line. This contextual method ensures that the original value is properly restored on exit and that it will not trigger additional recompute due to the temporary change of the discount value

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
[BOT] post-merge updates

dalonsod committed:
[IMP] sale_triple_discount: pre-commit auto fixes

dalonsod committed:
[IMP] sale_triple_discount: pre-commit auto fixes

dalonsod committed:
[MIG] sale_triple_discount: Migration to 17.0

dalonsod committed:
[MIG] sale_triple_discount: Migration to 17.0

sbejaoui committed:
[MIG] sale_triple_discount: migration 18.0

sbejaoui committed:
[MIG] sale_triple_discount: migration 18.0

sbejaoui committed:
[REF] sale_triple_discount: Consolidate discount in std field

inspired by https://github.com/OCA/account-invoicing/pull/1638

sbejaoui committed:
[REF] sale_triple_discount: Consolidate discount in std field

inspired by https://github.com/OCA/account-invoicing/pull/1638

sbejaoui committed:
[FIX] sale_triple_discount: hide discount_type and disable additive type

see https://github.com/OCA/sale-workflow/issues/3649

sbejaoui committed:
[FIX] sale_triple_discount: hide discount_type and disable additive type

see https://github.com/OCA/sale-workflow/issues/3649

sbejaoui committed:
[IMP] sale_triple_discount: add inverse for standard discount field

add an inverse to the standard discount field, which resets the three
discounts and sets discount1 with the value coming from the inverse.
this way, any standard process like pricelists will integrate with this
logic out of the box, without extra customization

sbejaoui committed:
[IMP] sale_triple_discount: add inverse for standard discount field

add an inverse to the standard discount field, which resets the three
discounts and sets discount1 with the value coming from the inverse.
this way, any standard process like pricelists will integrate with this
logic out of the box, without extra customization

sbiosca-s73 committed:
FIX sale_triple_discount: Calculate discounts when creating sales order lines

sbiosca-s73 committed:
FIX sale_triple_discount: Calculate discounts when creating sales order lines

lef-adhoc committed:
[DON'T MERGE] test-requirements.txt

ferran-S73 reviewed (approved):
Code review

lef-adhoc commented:
@sbejaoui Hi, could you check this PR?

lef-adhoc committed:
[IMP] sale_triple_discount: pre-commit execution

lef-adhoc committed:
[IMP] sale_triple_discount: pre-commit execution

lef-adhoc committed:
[MIG] sale_triple_discount: Migration to 19.0

lef-adhoc committed:
[MIG] sale_triple_discount: Migration to 19.0

dreispt commented:
/ocabot migration sale_triple_discount

Reyes4711-S73 reviewed (approved):
LGTM

Alexgars73 commented:
@lef-adhoc could you change the class text-right to text-end? text-right is obsolete and has no effect. Thanks!

alexey-pelykh commented on sale_triple_discount/models/sale_order_line.py:
```diff
@@ -0,0 +1,177 @@
+            raise ValidationError(
+                self.env._(
+                    "Sale order line %(name)s has unknown discounting type "
+                    "%(disc_type)s",
```
The `.pot` and `.po` files contain `%(dic_type)s` but this code passes `disc_type=` as the keyword argument. The translation template is out of sync with the source. Regenerate the `.pot` to fix; otherwise translated error messages will raise `KeyError` at runtime.

alexey-pelykh commented on sale_triple_discount/models/sale_order_line.py:
```diff
@@ -0,0 +1,177 @@
+        # calling super ensures the cache is coherent before applying the inverse
+        super()._compute_discount()
+        self._inverse_discount()
+        return True
```
Compute methods in Odoo 19.0 should not return a value. Remove the `return True`.

alexey-pelykh reviewed (commented):
## Migration Review

Good migration overall. The core discount logic, ORM patterns, view XML, and tests are solid. A few issues to address:

### Blocking

1. **Translation template mismatch** -- The `.pot` file (and all `.po` files) reference `%(dic_type)s` but the Python source uses `%(disc_type)s`. These are different placeholder names. The `.pot` needs regeneration to match the actual source string. Translations referencing the wrong placeholder will produce a `KeyError` at runtime.

2. **`_compute_discounts` returns `True`** -- Odoo 19.0 compute methods should not return a value. Returning `True` is a legacy pattern. Remove the `return True` statement.

### Non-blocking

3. **CI is red** -- All three test workflows are failing. The "Detect unreleased dependencies" failure is expected given the dependency on `OCA/account-invoicing#2144`. However, the actual test runs also fail, which should be investigated once the dependency PR is merged.

4. **`_additive_discount` clamps negatives to 0** -- The additive method returns 0 for negative sums, but the multiplicative method allows negative results (surcharges). This behavioral inconsistency could confuse users if additive mode is ever fully enabled. Since additive is currently blocked via `_discounting_type_additive_not_allowed`, this is informational only.

5. **`create` override fragility** -- The logic in `create` that conditionally calls `_compute_discounts` based on comparing `discount` and `discount1` values with `vals` is complex and could be fragile across Odoo upgrades. Consider adding a comment explaining the exact scenario this handles (e.g., when `discount` comes from a pricelist vs when it is manually set).

BhaveshHeliconia reviewed (commented):
`Additive` discount type missing due to open issue (https://github.com/OCA/sale-workflow/issues/3649), but LGTM for `Multiplicative` discount type

tao-thewarat reviewed (approved):
LGTM😎

lef-adhoc committed:
[FIX] sale_triple_discount: avoid crash saving SO with combo product

Making `discount` readonly (this module) makes the ORM drop it from the
create vals and recompute it during precompute (_add_precomputed_values).
For a combo item line the inherited _compute_discount calls
_get_linked_line().ensure_one() on a parent line that is not linked yet
during create -> "Expected singleton: sale.order.line()".

Defer combo items not linked to their parent yet in _compute_discount and
_compute_discounts; the module's create() recomputes them afterwards, once
native create() has set linked_line_id, so they inherit the parent's
discount. Add a regression test.

nicolascol commented:
Hi @alexey-pelykh 
Could you merge please?
